dev-docs/RFCs/v7.1/shader-module-injection-rfc.md

```diff
@@ -0,0 +1,63 @@
+# RFC: Shader Module Code Injection
+
+* **Author**: Tarek Sherif
+* **Date**: May, 2019
+* **Status**: **Draft**
+
+
+## Summary
+
+This RFC specifies a mechanism to allow shader modules to inject custom code into the source they are augmenting based on a defined set of lifecycle hook functions.
+
+
+## Background
+
+Shader modules currently only add variables and define functions in shaders. Without the ability to inject code in the `main` function, shaders that use them are forced modify their code to use the functionality that a shader module provides.
+
+There is a separate code-injection mechanism in, primarily used via `BaseModel`, that allows for code injection by the application. It limits injections to variable declaration segments, or the beginning or end of the `main` function. This leads to a relatively brittle system, where the application must write arbitrary GLSL code against shader code that might change at some point, and that it cannot affect any intermediate calculations.
+
+
+## Customers
+
+deck.gl's effects system is currently unscalable due to the fact that effects are implemented as shader modules and enabling effects currently requires that shaders in all layers be updated to call module functions.
+
+
+## Overview
+
+The foundation of the proposed system is a set of predifined hook functions into which modules can inject code, e.g:
+- `LUMAGL_fragmentColor(color)`
+- `LUMAGL_pickColor(color)`
+- `LUMAGL_viewMatrix(matrix)`
+- `LUMAGL_surfaceLight(surfaceColor, normal, eyePosition)`
+
+These functions will always be included in assembled shaders, and by default will be no-ops. Shader source code will use these functions to indicate that variables are ready for use, and modules can choose to inject code into these functions to modify the input data.
+
+It is important to note that these functions shouldn't be considered module-specific. They should generally be thought of as indicating moments in the lifecycle of a shader, the idea being that multiple modules can inject code into a particular function.
+
+
+## Example
+
+Setting the picking color of a fragment simply involves replacing the calculated fragment color with a provided picking color, if picking is enabled. Currently, this is done by including the picking module and directly calling the `picking_filterColor` function it provides. This could be done automatically by defining a hook function `LUMAGL_fragmentColor` to indicate that the fragment color has been finalized. In the fragment shader `main` function, this would appear as follows:
+```js
+void main() {
+  //...
+  LUMAGL_fragmentColor(gl_FragColor)
+}
+```
+
+The picking module would include an injection definition akin to the following:
+```js
+'LUMAGL_fragmentColor': 'color = picking_filterColor(color)'
```

Comment:
It is unclear to me how this applies to multiple injections.

For example, the picking module, the phong lighting module and the data filter module all need to filter the fragment color. The picking module must be applied last because nobody is supposed to modify the picking color. How does it work?

Reply:
Currently, injections are inserted in the order that their modules are included. So it's up to the application to include them in the right order.

Reply:
Yes, my original (half-baked) idea was that each module could define a relative priority for their injections (picking which always must come last could be `100` for instance). 

But in our discussion yesterday @tsherif convincingly argued that this can easily be added later and that we can start with module injection order for now.

Now the shader assembler does reorder modules to ensure that module dependency DAG is respected but that should not affect this order.